test/fixtures/s3-fixture/src/main/java/fixture/s3/S3HttpHandler.java

```diff
@@ -245,31 +245,49 @@ public void handle(final HttpExchange exchange) throws IOException {
-                        exchange.getResponseHeaders().add("Content-Type", "application/octet-stream");
-                        exchange.getResponseHeaders()
-                            .add("Content-Range", String.format(Locale.ROOT, "bytes %d-%d/%d", start, end, rangeBlob.length()));
-                        exchange.sendResponseHeaders(RestStatus.OK.getStatus(), rangeBlob.length());
-                        rangeBlob.writeTo(exchange.getResponseBody());
-                    }
-                } else {
+                if (blob == null) {
                     exchange.sendResponseHeaders(RestStatus.NOT_FOUND.getStatus(), -1);
+                    return;
                 }
+                final String range = exchange.getRequestHeaders().getFirst("Range");
+                if (range == null) {
+                    exchange.getResponseHeaders().add("Content-Type", "application/octet-stream");
+                    exchange.sendResponseHeaders(RestStatus.OK.getStatus(), blob.length());
+                    blob.writeTo(exchange.getResponseBody());
+                    return;
+                }
+
+                // S3 supports https://www.rfc-editor.org/rfc/rfc9110.html#name-range. The AWS SDK v1.x seems to always generate range
+                // requests with a header value like "Range: bytes=start-end" where both {@code start} and {@code end} are always defined
+                // (sometimes to very high value for {@code end}). It would be too tedious to fully support the RFC so S3HttpHandler only
+                // supports when both {@code start} and {@code end} are defined to match the SDK behavior.
+                final Matcher matcher = Pattern.compile("^bytes=([0-9]+)-([0-9]+)$").matcher(range);
+                if (matcher.matches() == false) {
+                    throw new AssertionError("Bytes range does not match expected pattern: " + range);
+                }
+                var groupStart = matcher.group(1);
+                var groupEnd = matcher.group(2);
+                if (groupStart == null || groupEnd == null) {
+                    throw new AssertionError("Bytes range does not match expected pattern: " + range);
+                }
+                int start = Integer.parseInt(groupStart);
+                int end = Integer.parseInt(groupEnd);
+                if (end < start) {
+                    exchange.getResponseHeaders().add("Content-Type", "application/octet-stream");
+                    exchange.sendResponseHeaders(RestStatus.OK.getStatus(), blob.length());
+                    blob.writeTo(exchange.getResponseBody());
+                    return;
+                } else if (blob.length() <= start) {
+                    exchange.getResponseHeaders().add("Content-Type", "application/octet-stream");
+                    exchange.sendResponseHeaders(RestStatus.REQUESTED_RANGE_NOT_SATISFIED.getStatus(), blob.length());
+                    return;
```

Comment:
Do we need to handle these cases? They seem kind of invalid to me. If so, could you add commentary explaining why?

Reply:
We don't need them for production code. But I've been surprised by the responses returned by the real S3 service in such cases and wanted them to be written somewhere.

I'll add a comment if that's OK to keep them.

Reply:
Sure; if this is how the real S3 behaves, could we cover it with a third-party test too?

Reply:
I added a 3rd party test specific for S3 [here](https://github.com/elastic/elasticsearch/pull/106239/files#diff-53a262822161aa5256f7e98bfbb00c52303541fb8603e92d8380675deab5c352R229) that covers the case of requesting a range that starts after the blob length.

The other case where S3 ignores requests with range end < start is maybe not worthly.